network/src/lib.rs

```diff
@@ -40,3 +46,35 @@ impl<T> std::future::Future for CancelHandler<T> {
         self.0.poll_unpin(cx).map(Result::unwrap)
     }
 }
+
+// This is the maximum number of network tasks that we will create for sending messages. It is a
+// limit per network struct - PrimaryNetwork, PrimaryToWorkerNetwork, and WorkerNetwork each have
+// their own limit.
+//
+// The exact number here probably isn't important, the key things is that it should be finite so
+// that we don't create unbounded numbers of tasks.
+pub const MAX_TASK_CONCURRENCY: usize = 200;
+
+async fn acquire_permit(semaphore: Arc<Semaphore>) -> Option<OwnedSemaphorePermit> {
+  if let Ok(permit) = semaphore.clone().try_acquire_owned() {
+    return Some(permit);
+  }
+
+  info!("concurrent task limit reached, waiting...");
+
+  semaphore.acquire_owned().await.ok()
+}
+
+pub async fn spawn_with_limited_concurrency(
+    semaphore: Arc<Semaphore>,
+    fut: impl Future<Output = ()> + Send + 'static,
+) -> JoinHandle<()> {
+    match acquire_permit(semaphore).await {
+        Some(permit) => tokio::spawn(async move {
+            let _permit = permit; // hold permit until future completes
+            fut.await;
+        }),
+        // semaphore has been closed, return an empty task and do nothing
+        None => tokio::spawn(async move {}),
+    }
+}
```

Comment:
The issue I have with this compared to e.g. the [BoundedExecutor](https://github.com/diem/diem/blob/416fdd3b18d02b2a099917467910d5a6865f2fed/common/bounded-executor/src/lib.rs) is the silent failure, which does not allow retry/failure management policies on top.

`BoundedExecutor` offers an alternative to the forcible queueing, thanks for looking at it.

Reply:
Well, BoundedExecutor just unwraps the permit result to avoid this issue